design/src/rfcs/rfc0022_error_context_and_compatibility.md

```diff
@@ -0,0 +1,370 @@
+properly display the error (for example, if the error has backtrace information, that would be
+inaccessible to the reporter).
+
+This situation might improve if the nightly `request_value`/`request_ref`/`provide` functions on
+`std::error::Error` are stabilized, since then contextual information needed for including things
+such as a backtrace could still be retrieved through the opaque error new-type.
+
+This RFC proposes that `Box<dyn Error + Send + Sync + 'static>` is used to refer to errors from libraries.
+An allocation to store this can be avoided if the error source is private, in which case, a reference to it
+can be returned from the `Error::source` implementation. Otherwise, if it's directly exposed as a public enum
+field, then it must be boxed.
+
+Ideally, errors should contain enough useful information on them that downcasting the underlying cause
+is not necessary. Customers needing to downcast the error source should be a last resort, and with the
+understanding that the type could change at a later date with no compile-time guarantees.
+
+Error Proposal
+--------------
+
+Taking a customer's perspective, there are two broad categories of errors:
+
+1. **Actionable:** Errors that can/should influence program flow
+2. **Informative:** Errors that inform that something went wrong, but where
+   it's not useful to match on the error to change program flow
+
+This RFC proposes that a consistent pattern be introduced to cover these two use cases for
+all errors in the public API for the Rust runtime crates and generated client crates.
+
+### Actionable error pattern
+
+Actionable errors must be represented as enums. If an error variant has an error source or additional
+contextual information, it must use a struct that is either embedded inside the enum, or separate where it makes
+sense to have private fields. For example:
+
+```rust
+pub enum Error {
+    // Good
+    #[non_exhaustive]
+    VariantA,
+
+    // Good
+    #[non_exhaustive]
+    VariantB {
+        some_additional_info: u32,
+        source: AnotherError // AnotherError is from this crate
+    },
+
+    // Good: This is exhaustive and uses a tuple, but its sole member is an extensible struct
+    VariantC(VariantC),
+
+    // Bad: Not ergonomic if additional context is added later
+    #[non_exhaustive]
+    VariantD(AnotherError),
+
+    // Bad: Exposes another library's error type in the public API,
+    // which makes upgrading or replacing that library a breaking change
+    #[non_exhaustive]
+    VariantF {
+        source: http::uri::InvalidUri
+    },
+
+    // Acceptable: Another library's error type may be in the box, and can be
+    // downcasted to, but the customer is then doing this downcasting at their
+    // own risk with no type safety guarantees that it will continue to work.
+    //
+    // Ideally, there would be a way to do opaque errors while playing
+    // nicely with fancy error reporters.
```

Comment:
I think you may be able to be more specific here but I'm not sure. I think I just have a bias against using the word "ideally". I'd rather see a definite statement.

When? Once an RFC is merged?
How? By extending it?